pennylane/tape/tape.py

```diff
@@ -872,6 +879,267 @@ def set_parameters(self, params, trainable_only=True):
+            )
+
+        shot_vector = device._shot_vector
+        if shot_vector is None:
+            if ret_type in (qml.operation.Expectation, qml.operation.Variance):
+
+                shape = (len(mps),)
+
+            elif ret_type == qml.operation.Probability:
+
+                wires_num_set = {len(meas.wires) for meas in mps}
+                same_num_wires = len(wires_num_set) == 1
+                if same_num_wires:
+                    # All probability measurements have the same number of
+                    # wires, gather the length from the first one
+
+                    len_wires = len(mps[0].wires)
+                    dim = QuantumTape._get_num_basis_states(len_wires, device)
+                    shape = (len(mps), dim)
+
+                else:
+                    # There are a varying number of wires that the probability
+                    # measurement processes act on
+                    shape = (sum(2 ** len(m.wires) for m in mps),)
+
+            elif ret_type == qml.operation.Sample:
+
+                shape = (len(mps), device.shots)
+
+        else:
+            # Shot vector was defined
+
+            if ret_type in (qml.operation.Expectation, qml.operation.Variance):
+                num = sum(shottup.copies for shottup in shot_vector)
+                shape = (num, len(mps))
+
+            elif ret_type == qml.operation.Probability:
+
+                wires_num_set = {len(meas.wires) for meas in mps}
+                same_num_wires = len(wires_num_set) == 1
+                if same_num_wires:
+                    # All probability measurements have the same number of
+                    # wires, gather the length from the first one
+
+                    len_wires = len(mps[0].wires)
+                    dim = QuantumTape._get_num_basis_states(len_wires, device)
+                    shot_copies_sum = sum(s.copies for s in shot_vector)
+                    shape = (shot_copies_sum, len(mps), dim)
+
+                else:
+                    # There are a varying number of wires that the probability
+                    # measurement processes act on
+                    # TODO: revisit when issues with this case are resolved
+                    raise UnsupportedTapeOperationError(
+                        "Getting the output shape of a tape with multiple probability measurements "
+                        "along with a device that defines a shot vector is not supported."
+                    )
+
+            elif ret_type == qml.operation.Sample:
+                shape = []
+                for shot_val in device.shot_vector:
+                    for _ in range(shot_val.copies):
+                        shots = shot_val.shots
+                        if shots != 1:
+                            shape.append(tuple([shots, len(mps)]))
+                        else:
+                            shape.append((len(mps),))
+
+        return shape
+
+    def get_output_shape(self, device):
```

Comment:
Do you think it is worth making the device optional, e.g., `device=None`? Since it is only used rarely (e.g., CV), and we should have a better long term approach that shouldn't require passing the device

Reply:
Yep, it's optional now.